File: python/list-ops/list_ops.py
def append(list1, list2):
    return concat([list1, list2])


def concat(lists):
    return [item for list in lists for item in list]


def filter(function, list):
    return [item for item in list if function(item)]


def length(list):
    return sum(1 for _ in list)


def map(function, list):
    return [function(item) for item in list]


def foldl(function, list, initial):
    return foldl(function, list[1:], function(initial, list[0])) \
        if length(list) > 0 else initial


def foldr(function, list, initial):
    return foldr(function, list[:-1], function(list[-1], initial)) \
        if length(list) > 0 else initial


def reverse(list):
    return list[::-1]

File: haskell/list-ops/src/ListOps.hs
module ListOps
  ( length
  , reverse
  , map
  , filter
  , foldr
  , foldl'
  , (++)
  , concat
  ) where

import           Prelude                 hiding ( (++)
                                                , concat
                                                , filter
                                                , foldr
                                                , length
                                                , map
                                                , reverse
                                                )

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' _ z []       = z
foldl' f z (x : xs) = z' `seq` foldl' f z' xs where z' = f z x

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr _ z []       = z
foldr f z (x : xs) = f x (foldr f z xs)

length :: [a] -> Int
length = foldr (\_ -> (1 +)) 0

reverse :: [a] -> [a]
reverse = foldl' (flip (:)) []

map :: (a -> b) -> [a] -> [b]
map f = foldr (\x z -> f x : z) []

filter :: (a -> Bool) -> [a] -> [a]
filter p = foldr (\x xs -> if p x then x : xs else xs) []

(++) :: [a] -> [a] -> [a]
xs ++ ys = foldr (:) ys xs

concat :: [[a]] -> [a]
concat = foldr (++) []
